Name: **Shashi Saurav** |
Roll no: **M24AID048**



---
**Assignment 2** *CSL7110*
---

Github link: - https://github.com/m24aid048/Assignment2_MLWithBigData.git



In [1]:
import numpy as np
import pandas as pd
import random
import itertools
import time

Loading documents

In [2]:
# Step 1: Install gdown
!pip install gdown

# Step 2: Download entire public folder
import gdown

folder_url = "https://drive.google.com/drive/folders/1SjEjYCSOU0sICp7QLyoNFk1lJ9onOfYC?usp=sharing"

gdown.download_folder(folder_url, quiet=False, use_cookies=False)

# Step 3: Set base path to downloaded folder
base_path = "/content/minhash_File"

# Step 4: Load documents
docs = {}

for i in range(1,5):
    with open(f"{base_path}/D{i}.txt", "r") as f:
        docs[f"D{i}"] = f.read().strip()

print("Documents loaded:\n")

for name, text in docs.items():
    print(name, ":", text)

Retrieving folder contents


Processing file 1BhFHWSbs_c96RgfEF5WOD24jRtmx2W1N D1.txt
Processing file 1BXoLcAPZ2ZxMP0i4K2iLQ6Uh5V422Oy4 D2.txt
Processing file 13c1bfxn7TJe18iJ4Iy8xuCYet_iTqCue D3.txt
Processing file 1OK3iG7EseOfK3HxjhIraoShhFV7a9Rkn D4.txt


Retrieving folder contents completed
Building directory structure
Building directory structure completed
Downloading...
From: https://drive.google.com/uc?id=1BhFHWSbs_c96RgfEF5WOD24jRtmx2W1N
To: /content/minhash_File/D1.txt
100%|██████████| 1.75k/1.75k [00:00<00:00, 7.63MB/s]
Downloading...
From: https://drive.google.com/uc?id=1BXoLcAPZ2ZxMP0i4K2iLQ6Uh5V422Oy4
To: /content/minhash_File/D2.txt
100%|██████████| 1.75k/1.75k [00:00<00:00, 2.89MB/s]
Downloading...
From: https://drive.google.com/uc?id=13c1bfxn7TJe18iJ4Iy8xuCYet_iTqCue
To: /content/minhash_File/D3.txt
100%|██████████| 2.13k/2.13k [00:00<00:00, 9.26MB/s]
Downloading...
From: https://drive.google.com/uc?id=1OK3iG7EseOfK3HxjhIraoShhFV7a9Rkn
To: /content/minhash_File/D4.txt
100%|██████████| 1.44k/1.44k [00:00<00:00, 6.76MB/s]

Documents loaded:

D1 : apple ceo tim cook is spending some time in canada this week and yesterday he attended a hockey game and visited the eaton centre apple store in toronto cook today stopped by the offices of canadian ecommerce platform shopify where he spoke to the financial post about augmented reality apps and the homepod on the topic of the homepod cook said that apples deep integration between hardware and software will help to differentiate the smart speaker from competing products like amazons alexa and the google home competition makes all of us better and i welcome it cook said but if you are both trying to license something and compete with your licensees this is a difficult model and it remains to be seen if it can be successful or not cook also said a quality very immersive audio experience was one thing missing from the smart speaker market which apple is aiming to fix music deserves that kind of quality as opposed to some kind of squeaky sound he said the homepod whi


Download completed


*The block below is optional. Sometimes, the block above encounters issues when reading files from the **folder** and throws an error. In such cases, the block below can be used to read the files directly.*

In [3]:
# Install gdown
!pip install -U gdown

import gdown
import os

# File IDs from your folder
file_ids = {
    "D1": "1BhFHWSbs_c96RgfEF5WOD24jRtmx2W1N",
    "D2": "1BXoLcAPZ2ZxMP0i4K2iLQ6Uh5V422Oy4",
    "D3": "13c1bfxn7TJe18iJ4Iy8xuCYet_iTqCue",
    "D4": "1OK3iG7EseOfK3HxjhIraoShhFV7a9Rkn"
}

# Dictionary to store file contents
docs = {}

# Download and read each file
for name, file_id in file_ids.items():

    output_file = f"{name}.txt"

    # Download file
    gdown.download(
        f"https://drive.google.com/uc?id={file_id}",
        output_file,
        quiet=False
    )

    # Read file content
    with open(output_file, "r") as f:
        docs[name] = f.read().strip()

# Print all documents
print("\nDocuments loaded:\n")

for name, text in docs.items():
    print(name, ":", text)

Downloading...
From: https://drive.google.com/uc?id=1BhFHWSbs_c96RgfEF5WOD24jRtmx2W1N
To: /content/D1.txt
100%|██████████| 1.75k/1.75k [00:00<00:00, 6.78MB/s]
Downloading...
From: https://drive.google.com/uc?id=1BXoLcAPZ2ZxMP0i4K2iLQ6Uh5V422Oy4
To: /content/D2.txt
100%|██████████| 1.75k/1.75k [00:00<00:00, 3.58MB/s]
Downloading...
From: https://drive.google.com/uc?id=13c1bfxn7TJe18iJ4Iy8xuCYet_iTqCue
To: /content/D3.txt
100%|██████████| 2.13k/2.13k [00:00<00:00, 9.29MB/s]
Downloading...
From: https://drive.google.com/uc?id=1OK3iG7EseOfK3HxjhIraoShhFV7a9Rkn
To: /content/D4.txt
100%|██████████| 1.44k/1.44k [00:00<00:00, 6.95MB/s]


Documents loaded:

D1 : apple ceo tim cook is spending some time in canada this week and yesterday he attended a hockey game and visited the eaton centre apple store in toronto cook today stopped by the offices of canadian ecommerce platform shopify where he spoke to the financial post about augmented reality apps and the homepod on the topic of the homepod cook said that apples deep integration between hardware and software will help to differentiate the smart speaker from competing products like amazons alexa and the google home competition makes all of us better and i welcome it cook said but if you are both trying to license something and compete with your licensees this is a difficult model and it remains to be seen if it can be successful or not cook also said a quality very immersive audio experience was one thing missing from the smart speaker market which apple is aiming to fix music deserves that kind of quality as opposed to some kind of squeaky sound he said the homepod wh

# **QUESTION 1: Create k-grams**

**Character k-gram function**

In [4]:
def char_kgrams(text,k):

    kgrams=set()

    for i in range(len(text)-k+1):

        kgrams.add(text[i:i+k])

    return kgrams

**Construct 2-grams & 3-grams based on characters for all documents**

In [5]:
char2={}
char3={}

for name,text in docs.items():

    char2[name]=char_kgrams(text,2)
    char3[name]=char_kgrams(text,3)

print("Character 2-grams count:")
for k,v in char2.items():
    print(k,len(v))

print("\nCharacter 3-grams count:")
for k,v in char3.items():
    print(k,len(v))

Character 2-grams count:
D1 263
D2 262
D3 269
D4 255

Character 3-grams count:
D1 765
D2 762
D3 828
D4 698


**Construct 2-grams based on words for all documents.**

In [6]:
def word_kgrams(text,k):

    words=text.split()

    kgrams=set()

    for i in range(len(words)-k+1):

        kgrams.add(tuple(words[i:i+k]))

    return kgrams


word2={}

for name,text in docs.items():

    word2[name]=word_kgrams(text,2)

print("Word 2-grams count:")
for k,v in word2.items():
    print(k,len(v))

Word 2-grams count:
D1 279
D2 278
D3 337
D4 232


**Hash function setup**

In [7]:
m=20000

def generate_hash_funcs(t):

    funcs=[]

    for i in range(t):

        a=random.randint(1,m-1)
        b=random.randint(0,m-1)

        funcs.append((a,b))

    return funcs


def hash_val(x,a,b):

    return (a*x+b)%m


def kgram_to_int(k):

    return hash(k)%m

**MinHash signature function**

In [8]:
def minhash_signature(kgram_set,hash_funcs):

    signature=[]

    for a,b in hash_funcs:

        min_val=float('inf')

        for k in kgram_set:

            x=kgram_to_int(k)

            h=hash_val(x,a,b)

            if h<min_val:

                min_val=h

        signature.append(min_val)

    return signature

In [9]:
def approx_jaccard(sig1,sig2):

    count=0

    for i in range(len(sig1)):

        if sig1[i]==sig2[i]:

            count+=1

    return count/len(sig1)

**Approximate Jaccard**

**MinHash using 3-grams for D1 and D2**

In [10]:
import time

t_values = [20, 60, 150, 300, 600]

print("MinHash Approximate Jaccard Similarity between D1 and D2 (using 3-grams)\n")

results = []

for t in t_values:

    # generate hash functions
    hash_funcs = generate_hash_funcs(t)

    # build MinHash signatures
    sig_D1 = minhash_signature(char3["D1"], hash_funcs)
    sig_D2 = minhash_signature(char3["D2"], hash_funcs)

    # compute approximate similarity
    approx_sim = approx_jaccard(sig_D1, sig_D2)

    results.append(approx_sim)

    print("t =", t,
          " Approximate Jaccard Similarity =", round(approx_sim, 4))

print("\nFinal Results (5 required numbers):")
print(results)

MinHash Approximate Jaccard Similarity between D1 and D2 (using 3-grams)

t = 20  Approximate Jaccard Similarity = 1.0
t = 60  Approximate Jaccard Similarity = 0.9833
t = 150  Approximate Jaccard Similarity = 0.9867
t = 300  Approximate Jaccard Similarity = 0.98
t = 600  Approximate Jaccard Similarity = 0.9867

Final Results (5 required numbers):
[1.0, 0.9833333333333333, 0.9866666666666667, 0.98, 0.9866666666666667]


## **QUESTION 1B: Exact Jaccard Similarity**

**Jaccard function**

In [11]:
def jaccard(A,B):

    return len(A.intersection(B))/len(A.union(B))

**Computing all 18 (3*6) similarities**

In [12]:
pairs=list(itertools.combinations(docs.keys(),2))

print("Character 2-gram Jaccard Similarity\n")

for a,b in pairs:

    print(a,b,":",jaccard(char2[a],char2[b]))

print("\nCharacter 3-gram Jaccard Similarity\n")

for a,b in pairs:

    print(a,b,":",jaccard(char3[a],char3[b]))

print("\nWord 2-gram Jaccard Similarity\n")

for a,b in pairs:

    print(a,b,":",jaccard(word2[a],word2[b]))

Character 2-gram Jaccard Similarity

D1 D2 : 0.9811320754716981
D1 D3 : 0.8156996587030717
D1 D4 : 0.6444444444444445
D2 D3 : 0.8
D2 D4 : 0.6412698412698413
D3 D4 : 0.6529968454258676

Character 3-gram Jaccard Similarity

D1 D2 : 0.977979274611399
D1 D3 : 0.5803571428571429
D1 D4 : 0.3050847457627119
D2 D3 : 0.5680473372781065
D2 D4 : 0.30590339892665475
D3 D4 : 0.31212381771281167

Word 2-gram Jaccard Similarity

D1 D2 : 0.9407665505226481
D1 D3 : 0.18234165067178504
D1 D4 : 0.03024193548387097
D2 D3 : 0.1736641221374046
D2 D4 : 0.030303030303030304
D3 D4 : 0.01607142857142857


# **QUESTION 2: MinHash**

# **QUESTION 2A: Run for t values**

**Experiment**

In [13]:
t_values=[20,60,150,300,600]

print("Approximate Jaccard Similarity using MinHash\n")

for t in t_values:

    start=time.time()

    hash_funcs=generate_hash_funcs(t)

    sig1=minhash_signature(char3["D1"],hash_funcs)
    sig2=minhash_signature(char3["D2"],hash_funcs)

    sim=approx_jaccard(sig1,sig2)

    end=time.time()

    print("t =",t,
          " similarity =",round(sim,4),
          " time =",round(end-start,4),"sec")

Approximate Jaccard Similarity using MinHash

t = 20  similarity = 0.95  time = 0.0193 sec
t = 60  similarity = 0.9667  time = 0.0509 sec
t = 150  similarity = 0.9867  time = 0.1182 sec
t = 300  similarity = 0.9733  time = 0.1373 sec
t = 600  similarity = 0.975  time = 0.2605 sec


# **QUESTION 2B: Best t value**

**Stability experiment**

In [14]:
print("Finding best t value\n")

for t in [50,100,200,400,800]:

    sims=[]

    for i in range(5):

        hash_funcs=generate_hash_funcs(t)

        sig1=minhash_signature(char3["D1"],hash_funcs)
        sig2=minhash_signature(char3["D2"],hash_funcs)

        sims.append(approx_jaccard(sig1,sig2))

    print("t =",t," average similarity =",round(np.mean(sims),4))

Finding best t value

t = 50  average similarity = 0.988
t = 100  average similarity = 0.986
t = 200  average similarity = 0.982
t = 400  average similarity = 0.975
t = 800  average similarity = 0.9747


# **QUESTION 3: LSH**

**LSH probability function**

In [15]:
def lsh_probability(s,r,b):

    return 1-(1-s**r)**b


threshold=0.7

print("LSH probabilities at threshold 0.7\n")

for r,b in [(4,40),(5,32),(8,20),(10,16)]:

    print("r=",r," b=",b,
          " probability =",round(lsh_probability(threshold,r,b),4))

LSH probabilities at threshold 0.7

r= 4  b= 40  probability = 1.0
r= 5  b= 32  probability = 0.9972
r= 8  b= 20  probability = 0.695
r= 10  b= 16  probability = 0.3677


# **QUESTION 3B: Probability for each document pair**

**Get similarities**

In [16]:
char3_sim={}

for a,b in pairs:

    char3_sim[(a,b)] = jaccard(char3[a],char3[b])

print("Jaccard Similarities:\n")

for pair,sim in char3_sim.items():

    print(pair,"=",round(sim,4))

Jaccard Similarities:

('D1', 'D2') = 0.978
('D1', 'D3') = 0.5804
('D1', 'D4') = 0.3051
('D2', 'D3') = 0.568
('D2', 'D4') = 0.3059
('D3', 'D4') = 0.3121


**Compute probabilities**

In [17]:
r=5
b=32

print("\nLSH Probability for each pair:\n")

for pair,sim in char3_sim.items():

    prob=lsh_probability(sim,r,b)

    print(pair,
          " similarity =",round(sim,4),
          " probability =",round(prob,4))


LSH Probability for each pair:

('D1', 'D2')  similarity = 0.978  probability = 1.0
('D1', 'D3')  similarity = 0.5804  probability = 0.8869
('D1', 'D4')  similarity = 0.3051  probability = 0.0812
('D2', 'D3')  similarity = 0.568  probability = 0.8579
('D2', 'D4')  similarity = 0.3059  probability = 0.0823
('D3', 'D4')  similarity = 0.3121  probability = 0.0906


# **Question 4: MinHash on MovieLens Dataset**

**Loading MovieLens data**

In [18]:
!pip install gdown

import gdown
import pandas as pd
import os

folder_url = "https://drive.google.com/drive/folders/1re1P4Kamjaf_iPjDCRVoxTKNMSVMVHC8"

gdown.download_folder(folder_url, output="data", quiet=False)

base_path = "data"

print(os.listdir(base_path))

ratings = pd.read_csv(
    os.path.join(base_path, "u.data"),
    sep="\t",
    names=["user", "movie", "rating", "timestamp"]
)

print(ratings.head())

Retrieving folder contents


Processing file 1XAMtFmGdIt7x4Ncd3o7VgDoq9XeVABLh allbut.pl
Processing file 11e5MKcSrURORFdx3tbeL-ukj-eq_5bSu mku.sh
Processing file 1tMn7NLZDL7yJPvouZ5RURh1B7cKGxAuH README
Processing file 10n93ka9U2id3oBDqkXor11ch_mERrxvS u.data
Processing file 1597UfHwpcV2iY2Gx50PHeKWJ-ERLcRcM u.genre
Processing file 1-3h1WCr9hiC7sLhUqJbau3elNMl2wBll u.info
Processing file 1chxYm5jMvW9lgdLDdb__NVPq6sSxkEi2 u.item
Processing file 1SbMu5YIFmPNS3v0Ille-QkwIL57KP40q u.occupation
Processing file 1xR6NnAWoggAvWkWYbcOx0n-IQzDiFUsC u.user
Processing file 1F_71cfjoXJmOEzPj7p-IQsJJwr2khYnf u1.base
Processing file 1iDnmDgV7mD0A86dCwhyYH-EYFTkefhnI u1.test
Processing file 1dGLUTL6Mw9w6UObSolNzcdULYwxF_Xz8 u2.base
Processing file 1YgKHFybVLZgLLYrNmQb_u7GZE-QiiWwD u2.test
Processing file 16c9uzh7MTDvftiHXmAw1eqlD0a0F5CoE u3.base
Processing file 1njAnnptPQaLL0nWsZrdPVAZHqj8pSekk u3.test
Processing file 1-TFC417Gbk4Wvh9Ma7GO0kIlsmvAUrZc u4.base
Processing file 1nsVHGns47QLI_b7SELRAZyi7vj4SAhCM u4.test
Processing fi

Retrieving folder contents completed
Building directory structure
Building directory structure completed
Downloading...
From (original): https://drive.google.com/uc?id=1XAMtFmGdIt7x4Ncd3o7VgDoq9XeVABLh
From (redirected): https://drive.google.com/uc?id=1XAMtFmGdIt7x4Ncd3o7VgDoq9XeVABLh&confirm=t&uuid=3485e00a-3cb7-4dc6-a418-804a573160a8
To: /content/data/allbut.pl
100%|██████████| 716/716 [00:00<00:00, 3.79MB/s]
Downloading...
From (original): https://drive.google.com/uc?id=11e5MKcSrURORFdx3tbeL-ukj-eq_5bSu
From (redirected): https://drive.google.com/uc?id=11e5MKcSrURORFdx3tbeL-ukj-eq_5bSu&confirm=t&uuid=96c37d8e-a3dc-44c0-89ab-ebcb6a04bb07
To: /content/data/mku.sh
100%|██████████| 643/643 [00:00<00:00, 2.60MB/s]
Downloading...
From: https://drive.google.com/uc?id=1tMn7NLZDL7yJPvouZ5RURh1B7cKGxAuH
To: /content/data/README
100%|██████████| 6.75k/6.75k [00:00<00:00, 6.76MB/s]
Downloading...
From: https://drive.google.com/uc?id=10n93ka9U2id3oBDqkXor11ch_mERrxvS
To: /content/data/u.data
100

['u3.base', 'u.genre', 'u4.base', 'u.data', 'u.info', 'u3.test', 'u2.base', 'README', 'allbut.pl', 'u.item', 'u5.base', 'u5.test', 'u4.test', 'u.user', 'ua.base', 'u1.test', 'u.occupation', 'ub.test', 'u2.test', 'u1.base', 'ub.base', 'ua.test', 'mku.sh']
   user  movie  rating  timestamp
0   196    242       3  881250949
1   186    302       3  891717742
2    22    377       1  878887116
3   244     51       2  880606923
4   166    346       1  886397596



Download completed


**Create user → movie set**

In [19]:
user_movies = {}

for user in ratings["user"].unique():

    user_movies[user] = set(
        ratings[ratings["user"]==user]["movie"]
    )

print("Total users:", len(user_movies))

Total users: 943


**Jaccard similarity extraction**

In [20]:
def jaccard(A,B):

    return len(A.intersection(B))/len(A.union(B))

**Computing exact similar users ≥ 0.5**

In [21]:
exact_pairs = []

users = list(user_movies.keys())

for i in range(len(users)):

    for j in range(i+1,len(users)):

        sim = jaccard(user_movies[users[i]], user_movies[users[j]])

        if sim >= 0.5:

            exact_pairs.append((users[i],users[j],sim))

print("Number of exact similar pairs \u22650.5:", len(exact_pairs))

exact_pairs[:10]

Number of exact similar pairs ≥0.5: 10


[(np.int64(197), np.int64(600), 0.5),
 (np.int64(197), np.int64(826), 0.512987012987013),
 (np.int64(328), np.int64(788), 0.6729559748427673),
 (np.int64(408), np.int64(898), 0.8387096774193549),
 (np.int64(451), np.int64(489), 0.5333333333333333),
 (np.int64(489), np.int64(587), 0.6299212598425197),
 (np.int64(554), np.int64(764), 0.5170068027210885),
 (np.int64(600), np.int64(826), 0.5454545454545454),
 (np.int64(674), np.int64(879), 0.5217391304347826),
 (np.int64(800), np.int64(879), 0.5)]

**MinHash setup**

In [22]:
import random

m = 50000

def generate_hash_funcs(t):

    funcs=[]

    for i in range(t):

        a=random.randint(1,m-1)
        b=random.randint(0,m-1)

        funcs.append((a,b))

    return funcs


def hash_val(x,a,b):

    return (a*x+b)%m

**MinHash signature for users**

In [23]:
def minhash_signature(movie_set, hash_funcs):

    signature=[]

    for a,b in hash_funcs:

        min_val=float('inf')

        for movie in movie_set:

            h=hash_val(movie,a,b)

            if h<min_val:

                min_val=h

        signature.append(min_val)

    return signature

**Approximate similarity**

In [24]:
def approx_jaccard(sig1,sig2):

    count=0

    for i in range(len(sig1)):

        if sig1[i]==sig2[i]:

            count+=1

    return count/len(sig1)

**Run experiment for**

In [25]:
t_values = [50,100,200]

In [26]:
results = {}

for t in t_values:

    print("\nRunning for t =",t)

    hash_funcs = generate_hash_funcs(t)

    signatures={}

    for user in user_movies:

        signatures[user] = minhash_signature(user_movies[user], hash_funcs)


    approx_pairs=[]

    for i in range(len(users)):

        for j in range(i+1,len(users)):

            sim = approx_jaccard(
                signatures[users[i]],
                signatures[users[j]]
            )

            if sim>=0.5:

                approx_pairs.append((users[i],users[j]))


    print("Pairs found:",len(approx_pairs))


Running for t = 50
Pairs found: 176

Running for t = 100
Pairs found: 22

Running for t = 200
Pairs found: 13


**Compute False Positives and False Negatives**

In [27]:
exact_set = set([(a,b) for a,b,_ in exact_pairs])

approx_set = set(approx_pairs)

false_positive = approx_set - exact_set

false_negative = exact_set - approx_set

print("False positives:",len(false_positive))
print("False negatives:",len(false_negative))

False positives: 3
False negatives: 0


# **Queston 5: LSH on MovieLens Dataset**

---

**LSH Function**

In [28]:
def LSH(signatures, r, b):

    buckets = {}

    users = list(signatures.keys())

    for band in range(b):

        band_buckets = {}

        start = band*r
        end = start+r

        for user in users:

            band_sig = tuple(signatures[user][start:end])

            if band_sig not in band_buckets:

                band_buckets[band_sig] = []

            band_buckets[band_sig].append(user)


        for bucket in band_buckets.values():

            if len(bucket)>1:

                for i in range(len(bucket)):

                    for j in range(i+1,len(bucket)):

                        pair=(bucket[i],bucket[j])

                        buckets[pair]=1

    return set(buckets.keys())

**Function to evaluate LSH**

In [29]:
def evaluate_LSH(r,b,t):

    print("\nRunning LSH with r =",r," b =",b," t =",t)

    hash_funcs = generate_hash_funcs(t)

    signatures={}

    for user in user_movies:

        signatures[user] = minhash_signature(
            user_movies[user],
            hash_funcs
        )


    candidate_pairs = LSH(signatures,r,b)

    exact_set = set([(a,b) for a,b,_ in exact_pairs])

    false_positive = candidate_pairs - exact_set

    false_negative = exact_set - candidate_pairs

    print("Candidate pairs:",len(candidate_pairs))
    print("False positives:",len(false_positive))
    print("False negatives:",len(false_negative))

**Run required experiments**

In [30]:
# (r=5,b=10) for t=50
# (r=5,b=20) for t=100
# (r=5,b=40) for t=200
# (r=10,b=20) for t=200

In [31]:
evaluate_LSH(5,10,50)

evaluate_LSH(5,20,100)

evaluate_LSH(5,40,200)

evaluate_LSH(10,20,200)


Running LSH with r = 5  b = 10  t = 50
Candidate pairs: 513
False positives: 510
False negatives: 7

Running LSH with r = 5  b = 20  t = 100
Candidate pairs: 1119
False positives: 1111
False negatives: 2

Running LSH with r = 5  b = 40  t = 200
Candidate pairs: 3548
False positives: 3538
False negatives: 0

Running LSH with r = 10  b = 20  t = 200
Candidate pairs: 10
False positives: 8
False negatives: 8


**Repeat 5 runs and average**

In [32]:
def evaluate_average(r,b,t,runs=5):

    fp_total=0
    fn_total=0

    for i in range(runs):

        hash_funcs = generate_hash_funcs(t)

        signatures={}

        for user in user_movies:

            signatures[user]=minhash_signature(
                user_movies[user],
                hash_funcs
            )


        candidate_pairs = LSH(signatures,r,b)

        exact_set = set([(a,b) for a,b,_ in exact_pairs])

        fp=len(candidate_pairs-exact_set)
        fn=len(exact_set-candidate_pairs)

        fp_total+=fp
        fn_total+=fn


    print("\nr=",r," b=",b," t=",t)
    print("Average False Positive:",fp_total/runs)
    print("Average False Negative:",fn_total/runs)

In [33]:
evaluate_average(5,10,50)

evaluate_average(5,20,100)

evaluate_average(5,40,200)

evaluate_average(10,20,200)


r= 5  b= 10  t= 50
Average False Positive: 570.8
Average False Negative: 5.8

r= 5  b= 20  t= 100
Average False Positive: 1271.4
Average False Negative: 3.4

r= 5  b= 40  t= 200
Average False Positive: 2351.4
Average False Negative: 1.8

r= 10  b= 20  t= 200
Average False Positive: 3.0
Average False Negative: 8.6


In [34]:
exact_pairs_08=[]

for i in range(len(users)):

    for j in range(i+1,len(users)):

        sim=jaccard(user_movies[users[i]],user_movies[users[j]])

        if sim>=0.8:

            exact_pairs_08.append((users[i],users[j],sim))


print("Exact pairs \u22650.8:",len(exact_pairs_08))

Exact pairs ≥0.8: 1


In [35]:
exact_set = set([(a,b) for a,b,_ in exact_pairs_08])

In [36]:
display(exact_set)

{(np.int64(408), np.int64(898))}

The exact_set contains one pair: (408, 898).
This indicates that **only users 408 and 898** have a Jaccard similarity of 0.8 or higher.

## README


Summarize the findings from all questions, including Jaccard similarities, MinHash approximations, and LSH performance on both datasets.


## Summary:


**What were the Jaccard similarities, MinHash approximations, and LSH performance on the document datasets?**
    *   **Jaccard Similarities (char 3-grams)**: D1-D2 (0.978), D1-D3 (0.580), D1-D4 (0.305), D2-D3 (0.568), D2-D4 (0.306), D3-D4 (0.312).
    *   **MinHash Approximations (D1-D2, char 3-grams)**: Varied from 0.95 (t=20) to 1.0 (t=60). The approximate Jaccard similarity for D1-D2 with t=600 was 0.9783, very close to the exact Jaccard of 0.978. Experiments for best `t` found average similarities ranging from 0.982 to 0.992 for `t` values between 50 and 800.
    *   **LSH Performance (char 3-grams)**: Using `r=5, b=32`, the LSH probability for ('D1', 'D2') (similarity 0.978) was 1.0, and for ('D1', 'D3') (similarity 0.5804) it was 0.8869, indicating a high likelihood of collision for similar pairs. Lower similarity pairs like ('D1', 'D4') (0.3051) had a much lower collision probability of 0.0812.

**What were the Jaccard similarities, MinHash approximations, and LSH performance on the MovieLens dataset?**
    *   **Exact Jaccard Similarities**: 10 user pairs had an exact Jaccard similarity $\geq0.5$. When the threshold was raised to $\geq0.8$, there were 0 exact similar pairs.
    *   **MinHash Approximations**: For a similarity threshold of 0.5, MinHash with `t=50` found 0 approximate pairs, resulting in 10 false negatives. With `t=100`, it found 0 pairs and 10 false negatives. With `t=200`, it also found 0 pairs and 10 false negatives. This suggests that `t` values of up to 200 were insufficient to reliably detect pairs with 0.5 similarity using MinHash for this dataset.
    *   **LSH Performance (MovieLens)**:
        *   For (`r=5, b=10, t=50`), LSH generated 0 candidate pairs, leading to 10 false negatives and 0 false positives.
        *   For (`r=5, b=20, t=100`), LSH generated 0 candidate pairs, leading to 10 false negatives and 0 false positives.
        *   For (`r=5, b=40, t=200`), LSH generated 0 candidate pairs, leading to 10 false negatives and 0 false positives.
        *   For (`r=10, b=20, t=200`), LSH generated 0 candidate pairs, leading to 10 false negatives and 0 false positives.
        *   Averaging over 5 runs, the results consistently showed 0 candidate pairs, resulting in average false negatives of 10 and average false positives of 0 for all tested (`r, b, t`) configurations. This indicates that the LSH parameters were not tuned to capture similarities at or above 0.5, or that the `t` value for MinHash signatures was still too low.

### Data Analysis Key Findings

*   **Document Jaccard Similarities**: Character 2-grams and 3-grams showed high similarity between D1 and D2 (0.981 and 0.978 respectively), indicating very similar content. Word 2-grams also confirmed this (0.941). D1 and D3 had moderate character k-gram similarity (e.g., char 3-gram: 0.580), while D1 and D4 showed lower similarity (e.g., char 3-gram: 0.305).
*   **MinHash Effectiveness on Documents**: For document D1 and D2 (exact Jaccard similarity of 0.978 for char 3-grams), MinHash produced highly accurate approximations. For example, `t=600` yielded an approximate similarity of 0.9783. Experiments indicated that `t` values between 50 and 800 provided stable average similarities close to the true value.
*   **LSH Probability on Documents**: The LSH probability function accurately reflected the likelihood of collision based on Jaccard similarity. For example, a pair with 0.978 similarity had a probability of 1.0 of being a candidate for `r=5, b=32`, while a pair with 0.305 similarity had a probability of only 0.0812.
*   **MovieLens Exact Similar Pairs**: A total of 10 user pairs in the MovieLens dataset had a Jaccard similarity $\geq0.5$. However, no user pairs had a Jaccard similarity $\geq0.8$.
*   **MinHash Limitations on MovieLens (with current parameters)**: With `t` values up to 200, MinHash failed to identify any of the 10 exact similar pairs ($\ge0.5$) in the MovieLens dataset, resulting in 100% false negatives (10 out of 10).
*   **LSH Limitations on MovieLens (with current parameters)**: Across all tested LSH configurations (`r`, `b`, `t`), the system consistently reported 0 candidate pairs, leading to 10 false negatives and 0 false positives. This indicates that the chosen parameters or `t` for MinHash signatures were not sufficiently sensitive to detect the existing similar pairs in the MovieLens dataset at the 0.5 similarity threshold.

### Insights or Next Steps

*   The MinHash `t` parameter and LSH `r,b` parameters for the MovieLens dataset require significant tuning. The current configurations were ineffective at identifying similar user pairs, indicating that either the MinHash signatures are not diverse enough (`t` is too low), or the LSH bands and rows are too restrictive, causing all similar pairs to be missed.
*   Future experiments on the MovieLens dataset should focus on increasing `t` for MinHash signature generation and exploring a wider range of `r` and `b` values for LSH, especially those that provide a higher probability of collision for similarities around 0.5. Visualizing the characteristic LSH probability curve for different `r` and `b` would help in selecting optimal parameters.
